### ISSUE TITLE
Retinanet anchor box implementation - question

### ISSUE THREAD

**Issue by pentanol2** [keras-team/keras-io#350]
In examples/vision/retinanet.py lines 257 and 258 why +0.5 was added ? 
```
rx = tf.range(feature_width, dtype=tf.float32) + 0.5
ry = tf.range(feature_height, dtype=tf.float32) + 0.5
```
Source:
https://github.com/keras-team/keras-io/blob/6d08fdc9ad469adb4c147c62b61b8e4af98bc9f7/examples/vision/retinanet.py#L257

**Comment by 8bitmp3** [comment]
I'd recommend changing the title to something like _"Retinanet anchor box implementation - question"_.

Maybe the following will also help to get a quicker answer.

> In examples/vision/retinanet.py lines 257 and 258 why +0.5 was added ?

Context (spelling it out should help):

```python
        rx = tf.range(feature_width, dtype=tf.float32) + 0.5
        ry = tf.range(feature_height, dtype=tf.float32) + 0.5
```

Source: https://github.com/keras-team/keras-io/blob/6d08fdc9ad469adb4c147c62b61b8e4af98bc9f7/examples/vision/retinanet.py#L257

More context:

```py
...
class AnchorBox:
    """Generates anchor boxes.
    This class has operations to generate anchor boxes for feature maps at
    strides `[8, 16, 32, 64, 128]`. Where each anchor each box is of the
    format `[x, y, width, height]`.
...
```py
    def _get_anchors(self, feature_height, feature_width, level):
        """Generates anchor boxes for a given feature map size and level
        Arguments:
          feature_height: An integer representing the height of the feature map.
          feature_width: An integer representing the width of the feature map.
          level: An integer representing the level of the feature map in the
            feature pyramid.
        Returns:
          anchor boxes with the shape
          `(feature_height * feature_width * num_anchors, 4)`
        """
        rx = tf.range(feature_width, dtype=tf.float32) + 0.5
        ry = tf.range(feature_height, dtype=tf.float32) + 0.5
        centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
        centers = tf.expand_dims(centers, axis=-2)
        centers = tf.tile(centers, [1, 1, self._num_anchors, 1])
        dims = tf.tile(
            self._anchor_dims[level - 3], [feature_height, feature_width, 1, 1]
        )
        anchors = tf.concat([centers, dims], axis=-1)
        return tf.reshape(
            anchors, [feature_height * feature_width * self._num_anchors, 4]
        )
```

**Comment by 8bitmp3** [comment]
> In examples/vision/retinanet.py lines 257 and 258 why +0.5 was added ?

My best "weak" guess is that "0.5" is the extra threshold that increases the bounding boxes around objects to make sure some things don't get missed by the detection algorithm. (Source: https://debuggercafe.com/object-detection-using-retinanet-with-pytorch-and-deep-learning/)

/assign @srihari-humbarwadi 

(cc @fchollet I noticed there are a few questions in Issuers about the RetinaNet example already)

**Comment by srihari-humbarwadi** [comment]
We place the anchor boxes at the centre of each of the cells of the grid, more about this [here](https://lilianweng.github.io/lil-log/2018/12/27/object-detection-part-4.html).  For a feature map of height and width hxw, the centres will be located at `[(0.5, 0.5), (1.5, 1.5) ...((h1-1)+0.5, (w-1)+0.5)]`. We then proceed to scale this by the stride of the feature map to get the absolute coordinates on the image, this is done by
 ````python 
rx = tf.range(feature_width, dtype=tf.float32) + 0.5
ry = tf.range(feature_height, dtype=tf.float32) + 0.5
centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
````

**Comment by pentanol2** [comment]
> We place the anchor boxes at the centre of each of the cells of the grid, more about this [here](https://lilianweng.github.io/lil-log/2018/12/27/object-detection-part-4.html). For a feature map of height and width hxw, the centres will be located at `[(0.5, 0.5), (1.5, 1.5) ...((h1-1)+0.5, (w-1)+0.5)]`. We then proceed to scale this by the stride of the feature map to get the absolute coordinates on the image, this is done by
> 
> ```python
> rx = tf.range(feature_width, dtype=tf.float32) + 0.5
> ry = tf.range(feature_height, dtype=tf.float32) + 0.5
> centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
> ```

You speak there about the feature height and width. Are you dealing with a feature map there or a grid cell ?
Or do you mean that the image is split into a grid, each grid is fed independently to the neural net and transformed to a set of feature maps? What do you mean exactly ?

**Comment by pentanol2** [comment]
> I'd recommend changing the title to something like _"Retinanet anchor box implementation - question"_.
> 
> Maybe the following will also help to get a quicker answer.
> 
> > In examples/vision/retinanet.py lines 257 and 258 why +0.5 was added ?
> 
> Context (spelling it out should help):
> 
> ```python
>         rx = tf.range(feature_width, dtype=tf.float32) + 0.5
>         ry = tf.range(feature_height, dtype=tf.float32) + 0.5
> ```
> 
> Source:
> 
> https://github.com/keras-team/keras-io/blob/6d08fdc9ad469adb4c147c62b61b8e4af98bc9f7/examples/vision/retinanet.py#L257
> 
> More context:
> 
> ```python
> ...
> class AnchorBox:
>     """Generates anchor boxes.
>     This class has operations to generate anchor boxes for feature maps at
>     strides `[8, 16, 32, 64, 128]`. Where each anchor each box is of the
>     format `[x, y, width, height]`.
> ...
> ```py
>     def _get_anchors(self, feature_height, feature_width, level):
>         """Generates anchor boxes for a given feature map size and level
>         Arguments:
>           feature_height: An integer representing the height of the feature map.
>           feature_width: An integer representing the width of the feature map.
>           level: An integer representing the level of the feature map in the
>             feature pyramid.
>         Returns:
>           anchor boxes with the shape
>           `(feature_height * feature_width * num_anchors, 4)`
>         """
>         rx = tf.range(feature_width, dtype=tf.float32) + 0.5
>         ry = tf.range(feature_height, dtype=tf.float32) + 0.5
>         centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
>         centers = tf.expand_dims(centers, axis=-2)
>         centers = tf.tile(centers, [1, 1, self._num_anchors, 1])
>         dims = tf.tile(
>             self._anchor_dims[level - 3], [feature_height, feature_width, 1, 1]
>         )
>         anchors = tf.concat([centers, dims], axis=-1)
>         return tf.reshape(
>             anchors, [feature_height * feature_width * self._num_anchors, 4]
>         )
> ```

which tensorflow version was used in implementing this code ?

**Comment by sachinprasadhs** [comment]
Hi,
It is always suggested to use latest Tensorflow/Keras for improved performances and bug fixes.
If your issue is resolved, could you please spare some time to close the issue.
For the latest update on the tutorial Retinanet, please follow the link here https://keras.io/examples/vision/retinanet/
